src/buildercore/bootstrap.py

```diff
@@ -164,24 +164,42 @@ def is_resourcing():
 
     stack_all_ec2_nodes(stackname, _setup_ec2_node, username=BOOTSTRAP_USER)
 
+def remove_topics_from_sqs_policy(policy, topic_arns):
+    """Removes statements from an SQS policy.
 
-def update_sqs_stack(stackname, **kwargs):
-    pdata = project_data_for_stackname(stackname)
-    if not pdata['aws']['sqs']:
-        return
-    context = context_handler.load_context(stackname)
-    setup_sqs(stackname, context.get('sqs', {}), pdata['aws']['region'])
+    These statements are created by boto's subscribe_sqs_queue()"""
 
-def unsub_sqs(stackname, context_sqs, region, dry_run=False):
+    def for_unsubbed_topic(statement):
+        return statement.get('Condition', {}).get('StringLike', {}).get('aws:SourceArn') in topic_arns
+
+    policy['Statement'] = list(filter(lambda s: not for_unsubbed_topic(s), policy['Statement']))
+    return policy
+
+def unsub_sqs(stackname, new_context, region, dry_run=False):
     sublist = core.all_sns_subscriptions(region, stackname)
-    struct = {}
-    for queue_name in context_sqs:
-        subscriptions = context_sqs[queue_name] # ll: [bus-articles--prod, ...]
-        struct[queue_name] = [sub for sub in sublist if sub['Topic'] not in subscriptions]
+
+    # compare project subscriptions to those actively subscribed to (sublist)
+    unsub_map = {}
+    permission_map = {}
+    for queue_name, subscriptions in new_context.items():
+        unsub_map[queue_name] = [sub for sub in sublist if sub['Topic'] not in subscriptions]
+        permission_map[queue_name] = [sub['TopicArn'] for sub in sublist if sub['Topic'] not in subscriptions]
+
     if not dry_run:
         sns = core.boto_sns_conn(region)
-        [sns.unsubscribe(sub['TopicArn']) for sub in utils.shallow_flatten(struct.values())]
-    return struct
+        for sub in utils.shallow_flatten(unsub_map.values()):
+            LOG.info("Unsubscribing %s from %s", sub['Topic'], stackname)
+            sns.unsubscribe(sub['SubscriptionArn'])
+        sqs = core.boto_sqs_conn(region)
+
+        for queue_name, topic_arns in permission_map.items():
+            # not an atomic update, but there's no other way to do it
```

Comment:
:+1:

Reply:
I wonder why the sqs policy isn't removed when unsubscribed? this is obviously what I was missing

Reply:
`Topic`, `Queue` and `Subscription` are three different objects. When unsubscribing, we delete the `Subscription` so things like `RawMessageDelivery that are attribute of the `Subscriptions` are removed as well. However, the `Policy` is a global property of the `Queue` so if you unsubscribe from just N out of M `Topics` it shouldn't be removed; rather, edited to remove references to these N topics but leaving alone the other statements referring to still-subscribed-to `Topics`.